File: src/world/chunk.ts
import { BlockType, BlockFaceTile } from "./block";

export const CHUNK_SIZE = 16;
const TOTAL = CHUNK_SIZE * CHUNK_SIZE * CHUNK_SIZE; // 4096
const N = CHUNK_SIZE;

/** Flat index from local (x, y, z) coordinates inside a chunk. */
function idx(x: number, y: number, z: number): number {
  return x + z * N + y * N * N;
}

// ── Greedy mesh face configurations ─────────────────────────────
//
// For each of the 6 axis-aligned face directions we define:
//   sliceAxis  – axis perpendicular to the face; we iterate slices along it
//   dim0       – first greedy expansion axis (→ U in local quad UV space)
//   dim1       – second greedy expansion axis (→ V in local quad UV space)
//   neighbor   – block-space offset to the cell that must be Air for the face
//                to be visible
//   normalSign – +1 → face plane is at s+1 on the slice axis; -1 → at s
//   light      – directional light multiplier (simple baked per-face shading)
//
// Face indices match BlockFaceTile order: [+Y, -Y, +X, -X, +Z, -Z]

interface SliceDef {
  faceIndex: number;
  sliceAxis: 0 | 1 | 2;
  dim0: 0 | 1 | 2;
  dim1: 0 | 1 | 2;
  neighbor: [number, number, number];
  normalSign: 1 | -1;
  light: number;
}

// ── Per-face tangent basis ─────────────────────────────────────
//
// Each row: [nx, ny, nz,  tx, ty, tz,  bitangentSign]
// where bitangent B = bitangentSign * cross(N, T).
//
// Derived from the exact corner winding used below for each face case so
// that TBN correctly maps the atlas _n texture to world/view space.
// Face order follows BlockFaceTile: [+Y, -Y, +X, -X, +Z, -Z].
const FACE_TBN: ReadonlyArray<
  readonly [number, number, number, number, number, number, number]
> = [
  [0, 1, 0, 1, 0, 0, -1], // 0: +Y
  [0, -1, 0, 1, 0, 0, 1], // 1: -Y
  [1, 0, 0, 0, 0, 1, -1], // 2: +X
  [-1, 0, 0, 0, 0, -1, -1], // 3: -X
  [0, 0, 1, 1, 0, 0, 1], // 4: +Z
  [0, 0, -1, -1, 0, 0, 1], // 5: -Z
] as const;

const SLICES: SliceDef[] = [
  {
    faceIndex: 0,
    sliceAxis: 1,
    dim0: 0,
    dim1: 2,
    neighbor: [0, 1, 0],
    normalSign: 1,
    light: 1.0,
  }, // +Y
  {
    faceIndex: 1,
    sliceAxis: 1,
    dim0: 0,
    dim1: 2,
    neighbor: [0, -1, 0],
    normalSign: -1,
    light: 0.5,
  }, // -Y
  {
    faceIndex: 2,
    sliceAxis: 0,
    dim0: 2,
    dim1: 1,
    neighbor: [1, 0, 0],
    normalSign: 1,
    light: 0.8,
  }, // +X
  {
    faceIndex: 3,
    sliceAxis: 0,
    dim0: 2,
    dim1: 1,
    neighbor: [-1, 0, 0],
    normalSign: -1,
    light: 0.8,
  }, // -X
  {
    faceIndex: 4,
    sliceAxis: 2,
    dim0: 0,
    dim1: 1,
    neighbor: [0, 0, 1],
    normalSign: 1,
    light: 0.7,
  }, // +Z
  {
    faceIndex: 5,
    sliceAxis: 2,
    dim0: 0,
    dim1: 1,
    neighbor: [0, 0, -1],
    normalSign: -1,
    light: 0.7,
  }, // -Z
];

export interface ChunkMesh {
  positions: Float32Array;
  /**
   * Per-vertex data packed as vec4: [localU, localV, tileIndex, light].
   * localU/V span 0 → (quad width/height) so the vertex shader can tile
   * the atlas sprite across the entire merged quad using fract().
   * `light` is the per-face directional shading factor (0.5 bottom – 1.0 top).
   */
  uvls: Float32Array;
  /** vec3 per vertex – object-space face normal (one of the 6 axis directions). */
  normals: Float32Array;
  /**
   * vec4 per vertex – tangent frame for normal mapping.
   *   xyz = object-space tangent direction (aligned with atlas U axis)
   *   w   = bitangent sign: B = w * cross(N, T)
   */
  tangents: Float32Array;
  /**
   * vec2 per vertex – lightmap values (sky, block), both normalised to [0, 1].
   *   x = skyLight  / 15  — how much sky light reaches this face
   *   y = blockLight / 15 — how much block-source light reaches this face
   */
  lights: Float32Array;
  vertexCount: number;
}

export interface ChunkSnapshot {
  cx: number;
  cz: number;
  blocks: Uint8Array;
}

export class Chunk {
  /** Block IDs stored in a flat array of length 4096. */
  blocks = new Uint8Array(TOTAL);

  /**
   * Per-block sky light level (0-15).
   * 15 = directly under open sky; propagated BFS horizontally.
   * Computed by recomputeWorldLight() in light-engine.ts.
   */
  skyLight = new Uint8Array(TOTAL);

  /**
   * Per-block emissive / torch light level (0-15).
   * Starts from BLOCK_EMISSION seeds and is BFS-propagated outward.
   * Computed by recomputeWorldLight() in light-engine.ts.
   */
  blockLight = new Uint8Array(TOTAL);

  /** Chunk coordinate in world-chunk space (multiply by 16 for world pos). */
  cx: number;
  cz: number;

  // GPU handles (assigned after upload)
  posBuffer: WebGLBuffer | null = null;
  uvBuffer: WebGLBuffer | null = null;
  vertexCount = 0;

  constructor(cx: number, cz: number) {
    this.cx = cx;
    this.cz = cz;
  }

  getBlock(x: number, y: number, z: number): BlockType {
    if (x < 0 || x >= N || y < 0 || y >= N || z < 0 || z >= N)
      return BlockType.Air;
    return this.blocks[idx(x, y, z)] as BlockType;
  }

  setBlock(x: number, y: number, z: number, type: BlockType): void {
    this.blocks[idx(x, y, z)] = type;
  }

  /**
   * Returns the sky light at (x, y, z).
   * Out-of-bounds → 15 (above world = open sky; below/side = no sky).
   */
  getSkyLight(x: number, y: number, z: number): number {
    if (y >= N) return 15; // above world = full sky
    if (x < 0 || x >= N || y < 0 || z < 0 || z >= N) return 0;
    return this.skyLight[idx(x, y, z)];
  }

  /**
   * Returns the block light at (x, y, z).
   * Out-of-bounds → 0 (no emissive source outside loaded chunk).
   */
  getBlockLight(x: number, y: number, z: number): number {
    if (x < 0 || x >= N || y < 0 || y >= N || z < 0 || z >= N) return 0;
    return this.blockLight[idx(x, y, z)];
  }

  toSnapshot(): ChunkSnapshot {
    return {
      cx: this.cx,
      cz: this.cz,
      blocks: new Uint8Array(this.blocks),
    };
  }

  static fromSnapshot(snapshot: ChunkSnapshot): Chunk {
    const chunk = new Chunk(snapshot.cx, snapshot.cz);
    if (snapshot.blocks.length !== TOTAL) {
      throw new Error(
        `Invalid chunk snapshot size: expected ${TOTAL}, got ${snapshot.blocks.length}`,
      );
    }
    chunk.blocks.set(snapshot.blocks);
    return chunk;
  }

  /**
   * Greedy mesh builder (Mikola Lysenko algorithm).
   *
   * For each of the 6 face directions we sweep every 16-block slice
   * perpendicular to the face normal:
   *   1. Build a 16×16 visibility mask: each cell holds an encoded value
   *      combining the atlas tile index and per-face light levels, or −1
   *      if the face is hidden.
   *   2. Greedily merge adjacent cells with the same encoded value into the
   *      largest axis-aligned rectangle, marking merged cells as used.
   *   3. Emit one quad (6 vertices / 2 triangles) per rectangle.
   *
   * Light values are sampled at the air block adjacent to each face (the
   * outward-facing neighbor). This ensures that a cave face lit by a torch
   * in the next block correctly picks up that block's light.
   *
   * Merging only happens when tile index AND light levels are equal, so
   * a gradient of sky light across a cliff face is never collapsed into a
   * single incorrectly-lit quad.
   */
  buildMesh(): ChunkMesh {
    const positions: number[] = [];
    const uvls: number[] = []; // 4 components per vertex: [localU, localV, tileIndex, light]
    const normals: number[] = []; // 3 components per vertex
    const tangents: number[] = []; // 4 components per vertex (xyz + bitangent sign)
    const lights: number[] = []; // 2 components per vertex: [skyL/15, blockL/15]

    // Reusable scratch arrays — allocated once to avoid per-slice GC pressure.
    //
    // mask encoding (Int32Array, value ≥ 0 = visible face):
    //   bits  0- 3: tileIdx  (0-11, 4 bits)
    //   bits  4- 7: skyLight  (0-15, 4 bits) — from air-side neighbor
    //   bits  8-11: blockLight (0-15, 4 bits) — from air-side neighbor
    //   = max encoded value: 11 | (15 << 4) | (15 << 8) = 0xBFF = 3071  (fits in Int16)
    //   -1 = face not visible
    const mask = new Int32Array(N * N);
    const used = new Uint8Array(N * N);
    const coord = [0, 0, 0];

    for (const sl of SLICES) {
      const { faceIndex, sliceAxis, dim0, dim1, neighbor, normalSign, light } =
        sl;

      for (let s = 0; s < N; s++) {
        // ── 1. Build visibility mask for this slice ─────────────
        mask.fill(-1);
        for (let j = 0; j < N; j++) {
          for (let i = 0; i < N; i++) {
            coord[sliceAxis] = s;
            coord[dim0] = i;
            coord[dim1] = j;
            const block = this.getBlock(coord[0], coord[1], coord[2]);
            if (block === BlockType.Air) continue;

            // Check the block on the outward side — must be Air.
            const nx = coord[0] + neighbor[0];
            const ny = coord[1] + neighbor[1];
            const nz = coord[2] + neighbor[2];
            if (this.getBlock(nx, ny, nz) !== BlockType.Air) continue;

            // Sample light at the air-side neighbor block.
            const skyL = this.getSkyLight(nx, ny, nz);
            const blockL = this.getBlockLight(nx, ny, nz);

            const faceTiles = BlockFaceTile[block] ?? [3, 3, 3, 3, 3, 3];
            const tileIdx = faceTiles[faceIndex];
            mask[i + j * N] = tileIdx | (skyL << 4) | (blockL << 8);
          }
        }

        // ── 2 & 3. Greedy sweep → emit merged quads ─────────────
        used.fill(0);
        for (let j = 0; j < N; j++) {
          for (let i = 0; i < N; i++) {
            if (used[i + j * N]) continue;
            const encoded = mask[i + j * N];
            if (encoded < 0) continue;

            // Grow width along dim0 (i direction).
            let w = 1;
            while (
              i + w < N &&
              !used[i + w + j * N] &&
              mask[i + w + j * N] === encoded
            )
              w++;

            // Grow height along dim1 (j direction).
            let h = 1;
            expand: while (j + h < N) {
              for (let k = i; k < i + w; k++) {
                if (used[k + (j + h) * N] || mask[k + (j + h) * N] !== encoded)
                  break expand;
              }
              h++;
            }

            // Mark merged cells as used.
            for (let dj = 0; dj < h; dj++)
              for (let di = 0; di < w; di++) used[i + di + (j + dj) * N] = 1;

            // Decode tile and light from the encoded mask value.
            const tileIdx = encoded & 0xf;
            const skyL = (encoded >> 4) & 0xf;
            const blockL = (encoded >> 8) & 0xf;

            // ── Emit the merged quad ──────────────────────────────
            // Face plane sits at sv on the slice axis.
            const sv = normalSign > 0 ? s + 1 : s;

            // Build a position from (slicePos, dim0Pos, dim1Pos).
            type V3 = [number, number, number];
            const p = (sa: number, da: number, db: number): V3 => {
              const c: V3 = [0, 0, 0];
              c[sliceAxis] = sa;
              c[dim0] = da;
              c[dim1] = db;
              return c;
            };

            // Four corners with CCW winding (viewed from outside the face).
            // UV pattern A (U→dim0, V→dim1): faces +Y, +X
            // UV pattern B (U→dim0, V→dim1, different winding): faces -Y, +Z
            // Reversed-U variants for -X and -Z preserve correct texture orientation.
            type Corner = [V3, [number, number]];
            let corners: Corner[];
            switch (faceIndex) {
              case 0: // +Y  dim0=X(i), dim1=Z(j) — pattern A
                corners = [
                  [p(sv, i, j), [0, 0]],
                  [p(sv, i, j + h), [0, h]],
                  [p(sv, i + w, j + h), [w, h]],
                  [p(sv, i + w, j), [w, 0]],
                ];
                break;
              case 1: // -Y  dim0=X(i), dim1=Z(j) — pattern B
                corners = [
                  [p(sv, i, j), [0, 0]],
                  [p(sv, i + w, j), [w, 0]],
                  [p(sv, i + w, j + h), [w, h]],
                  [p(sv, i, j + h), [0, h]],
                ];
                break;
              case 2: // +X  dim0=Z(i), dim1=Y(j) — pattern A
                corners = [
                  [p(sv, i, j), [0, 0]],
                  [p(sv, i, j + h), [0, h]],
                  [p(sv, i + w, j + h), [w, h]],
                  [p(sv, i + w, j), [w, 0]],
                ];
                break;
              case 3: // -X  dim0=Z(i), dim1=Y(j) — reversed U along dim0
                corners = [
                  [p(sv, i + w, j), [0, 0]],
                  [p(sv, i + w, j + h), [0, h]],
                  [p(sv, i, j + h), [w, h]],
                  [p(sv, i, j), [w, 0]],
                ];
                break;
              case 4: // +Z  dim0=X(i), dim1=Y(j) — pattern B
                corners = [
                  [p(sv, i, j), [0, 0]],
                  [p(sv, i + w, j), [w, 0]],
                  [p(sv, i + w, j + h), [w, h]],
                  [p(sv, i, j + h), [0, h]],
                ];
                break;
              default: // -Z  dim0=X(i), dim1=Y(j) — reversed U along dim0
                corners = [
                  [p(sv, i + w, j), [0, 0]],
                  [p(sv, i, j), [w, 0]],
                  [p(sv, i, j + h), [w, h]],
                  [p(sv, i + w, j + h), [0, h]],
                ];
                break;
            }

            // Two triangles: (0,1,2) and (0,2,3).
            const tbn = FACE_TBN[faceIndex];
            const skyN = skyL / 15;
            const blockN = blockL / 15;
            for (const vi of [0, 1, 2, 0, 2, 3]) {
              const [pos, uv] = corners[vi];
              positions.push(pos[0], pos[1], pos[2]);
              uvls.push(uv[0], uv[1], tileIdx, light);
              normals.push(tbn[0], tbn[1], tbn[2]);
              tangents.push(tbn[3], tbn[4], tbn[5], tbn[6]);
              lights.push(skyN, blockN);
            }
          }
        }
      }
    }

    return {
      positions: new Float32Array(positions),
      uvls: new Float32Array(uvls),
      normals: new Float32Array(normals),
      tangents: new Float32Array(tangents),
      lights: new Float32Array(lights),
      vertexCount: positions.length / 3,
    };
  }
}

File: game/src/world/chunk.rs
use crate::world::block::{block_face_tiles, is_opaque, BlockType};

pub const CHUNK_SIZE: usize = 16;
const TOTAL: usize = CHUNK_SIZE * CHUNK_SIZE * CHUNK_SIZE; // 4096
const N: usize = CHUNK_SIZE;

#[inline]
fn idx(x: usize, y: usize, z: usize) -> usize {
    x + z * N + y * N * N
}

// ── Greedy mesh face definitions ─────────────────────

struct SliceDef {
    face_index: usize,
    slice_axis: usize,
    dim0: usize,
    dim1: usize,
    neighbor: [i32; 3],
    normal_sign: i32,
    light: f32,
}

/// Per-face TBN: [nx, ny, nz, tx, ty, tz, bitangent_sign]
/// Face order: [+Y, -Y, +X, -X, +Z, -Z]
const FACE_TBN: [[f32; 7]; 6] = [
    [0.0, 1.0, 0.0, 1.0, 0.0, 0.0, -1.0],   // +Y
    [0.0, -1.0, 0.0, 1.0, 0.0, 0.0, 1.0],   // -Y
    [1.0, 0.0, 0.0, 0.0, 0.0, 1.0, -1.0],   // +X
    [-1.0, 0.0, 0.0, 0.0, 0.0, -1.0, -1.0], // -X
    [0.0, 0.0, 1.0, 1.0, 0.0, 0.0, 1.0],    // +Z
    [0.0, 0.0, -1.0, -1.0, 0.0, 0.0, 1.0],  // -Z
];

const SLICES: [SliceDef; 6] = [
    SliceDef {
        face_index: 0,
        slice_axis: 1,
        dim0: 0,
        dim1: 2,
        neighbor: [0, 1, 0],
        normal_sign: 1,
        light: 1.0,
    }, // +Y
    SliceDef {
        face_index: 1,
        slice_axis: 1,
        dim0: 0,
        dim1: 2,
        neighbor: [0, -1, 0],
        normal_sign: -1,
        light: 0.5,
    }, // -Y
    SliceDef {
        face_index: 2,
        slice_axis: 0,
        dim0: 2,
        dim1: 1,
        neighbor: [1, 0, 0],
        normal_sign: 1,
        light: 0.8,
    }, // +X
    SliceDef {
        face_index: 3,
        slice_axis: 0,
        dim0: 2,
        dim1: 1,
        neighbor: [-1, 0, 0],
        normal_sign: -1,
        light: 0.8,
    }, // -X
    SliceDef {
        face_index: 4,
        slice_axis: 2,
        dim0: 0,
        dim1: 1,
        neighbor: [0, 0, 1],
        normal_sign: 1,
        light: 0.7,
    }, // +Z
    SliceDef {
        face_index: 5,
        slice_axis: 2,
        dim0: 0,
        dim1: 1,
        neighbor: [0, 0, -1],
        normal_sign: -1,
        light: 0.7,
    }, // -Z
];

// ── Mesh output ──────────────────────────────────────

pub struct ChunkMesh {
    pub positions: Vec<f32>,
    pub normals: Vec<f32>,
    /// [localU, localV, tileIndex, light] per vertex
    pub uvls: Vec<f32>,
    /// [tx, ty, tz, bitangent_sign] per vertex
    pub tangents: Vec<f32>,
    /// [skyLight/15, blockLight/15] per vertex
    pub lights: Vec<f32>,
    pub vertex_count: u32,
}

pub struct ChunkSnapshot {
    pub cx: i32,
    pub cz: i32,
    pub blocks: Vec<u8>,
}

// ── Light sampling helper (world-aware) ──────────────

/// Callback used by the mesh builder to sample light at
/// a world-space position (for cross-chunk sampling).
pub trait LightSampler {
    fn sky_light(&self, wx: i32, wy: i32, wz: i32) -> u8;
    fn block_light(&self, wx: i32, wy: i32, wz: i32) -> u8;
}

// ── Chunk ─────────────────────────────────────────────

#[derive(Clone)]
pub struct Chunk {
    pub blocks: Vec<u8>,
    pub sky_light: Vec<u8>,
    pub block_light: Vec<u8>,
    pub cx: i32,
    pub cz: i32,
}

impl Chunk {
    pub fn new(cx: i32, cz: i32) -> Self {
        Self {
            blocks: vec![0u8; TOTAL],
            sky_light: vec![0u8; TOTAL],
            block_light: vec![0u8; TOTAL],
            cx,
            cz,
        }
    }

    pub fn get_block(&self, x: i32, y: i32, z: i32) -> BlockType {
        if x < 0 || x >= N as i32 || y < 0 || y >= N as i32 || z < 0 || z >= N as i32 {
            return BlockType::Air;
        }
        BlockType::from_u8(self.blocks[idx(x as usize, y as usize, z as usize)])
    }

    pub fn set_block(&mut self, x: usize, y: usize, z: usize, t: BlockType) {
        self.blocks[idx(x, y, z)] = t as u8;
    }

    pub fn get_sky_light(&self, x: i32, y: i32, z: i32) -> u8 {
        if y >= N as i32 {
            return 15;
        }
        if x < 0 || x >= N as i32 || y < 0 || z < 0 || z >= N as i32 {
            return 0;
        }
        self.sky_light[idx(x as usize, y as usize, z as usize)]
    }

    pub fn get_block_light(&self, x: i32, y: i32, z: i32) -> u8 {
        if x < 0 || x >= N as i32 || y < 0 || y >= N as i32 || z < 0 || z >= N as i32 {
            return 0;
        }
        self.block_light[idx(x as usize, y as usize, z as usize)]
    }

    pub fn to_snapshot(&self) -> ChunkSnapshot {
        ChunkSnapshot {
            cx: self.cx,
            cz: self.cz,
            blocks: self.blocks.clone(),
        }
    }

    pub fn from_snapshot(s: ChunkSnapshot) -> Self {
        let mut c = Self::new(s.cx, s.cz);
        c.blocks = s.blocks;
        c
    }

    /// Build a greedy mesh. Pass a `LightSampler` (the
    /// World) for correct cross-chunk boundary lighting.
    pub fn build_mesh(&self, sampler: Option<&dyn LightSampler>) -> ChunkMesh {
        let mut positions: Vec<f32> = Vec::new();
        let mut uvls: Vec<f32> = Vec::new();
        let mut normals: Vec<f32> = Vec::new();
        let mut tangents: Vec<f32> = Vec::new();
        let mut lights: Vec<f32> = Vec::new();

        let mut mask = vec![-1i32; N * N];
        let mut used = vec![0u8; N * N];
        let mut coord = [0i32; 3];

        for sl in &SLICES {
            let SliceDef {
                face_index,
                slice_axis,
                dim0,
                dim1,
                neighbor,
                normal_sign,
                light,
            } = sl;

            for s in 0..N {
                // ── 1. Build visibility mask ──────────────
                mask.fill(-1);
                for j in 0..N {
                    for i in 0..N {
                        coord[*slice_axis] = s as i32;
                        coord[*dim0] = i as i32;
                        coord[*dim1] = j as i32;

                        let block = self.get_block(coord[0], coord[1], coord[2]);
                        if block == BlockType::Air {
                            continue;
                        }

                        let nx = coord[0] + neighbor[0];
                        let ny = coord[1] + neighbor[1];
                        let nz = coord[2] + neighbor[2];
                        if self.get_block(nx, ny, nz) != BlockType::Air {
                            continue;
                        }

                        // Sample light at air-side neighbor
                        let (sky_l, block_l) = if let Some(ws) = sampler {
                            let wx = self.cx * N as i32 + nx;
                            let wz = self.cz * N as i32 + nz;
                            (ws.sky_light(wx, ny, wz), ws.block_light(wx, ny, wz))
                        } else {
                            (
                                self.get_sky_light(nx, ny, nz),
                                self.get_block_light(nx, ny, nz),
                            )
                        };

                        let tiles = block_face_tiles(block);
                        let tile_idx = tiles[*face_index] as i32;
                        mask[i + j * N] =
                            tile_idx | ((sky_l as i32) << 4) | ((block_l as i32) << 8);
                    }
                }

                // ── 2&3. Greedy sweep → emit quads ───────
                used.fill(0);
                for j in 0..N {
                    for i in 0..N {
                        if used[i + j * N] != 0 {
                            continue;
                        }
                        let encoded = mask[i + j * N];
                        if encoded < 0 {
                            continue;
                        }

                        // Grow width along dim0
                        let mut w = 1usize;
                        while i + w < N
                            && used[i + w + j * N] == 0
                            && mask[i + w + j * N] == encoded
                        {
                            w += 1;
                        }

                        // Grow height along dim1
                        let mut h = 1usize;
                        'expand: while j + h < N {
                            for k in i..i + w {
                                if used[k + (j + h) * N] != 0 || mask[k + (j + h) * N] != encoded {
                                    break 'expand;
                                }
                            }
                            h += 1;
                        }

                        // Mark used
                        for dj in 0..h {
                            for di in 0..w {
                                used[i + di + (j + dj) * N] = 1;
                            }
                        }

                        let tile_idx = (encoded & 0xf) as f32;
                        let sky_l = ((encoded >> 4) & 0xf) as f32;
                        let block_l = ((encoded >> 8) & 0xf) as f32;

                        let sv = if *normal_sign > 0 {
                            (s + 1) as i32
                        } else {
                            s as i32
                        };

                        // Helper to build a position
                        let p = |sa: i32, da: i32, db: i32| -> [f32; 3] {
                            let mut c = [0f32; 3];
                            c[*slice_axis] = sa as f32;
                            c[*dim0] = da as f32;
                            c[*dim1] = db as f32;
                            c
                        };

                        let ii = i as i32;
                        let jj = j as i32;
                        let ww = w as i32;
                        let hh = h as i32;

                        // Four corners per face index
                        // (pos, [u, v])
                        let corners: [([f32; 3], [f32; 2]); 4] = match face_index {
                            0 => [
                                // +Y
                                (p(sv, ii, jj), [0.0, 0.0]),
                                (p(sv, ii, jj + hh), [0.0, hh as f32]),
                                (p(sv, ii + ww, jj + hh), [ww as f32, hh as f32]),
                                (p(sv, ii + ww, jj), [ww as f32, 0.0]),
                            ],
                            1 => [
                                // -Y
                                (p(sv, ii, jj), [0.0, 0.0]),
                                (p(sv, ii + ww, jj), [ww as f32, 0.0]),
                                (p(sv, ii + ww, jj + hh), [ww as f32, hh as f32]),
                                (p(sv, ii, jj + hh), [0.0, hh as f32]),
                            ],
                            2 => [
                                // +X
                                (p(sv, ii, jj), [0.0, 0.0]),
                                (p(sv, ii, jj + hh), [0.0, hh as f32]),
                                (p(sv, ii + ww, jj + hh), [ww as f32, hh as f32]),
                                (p(sv, ii + ww, jj), [ww as f32, 0.0]),
                            ],
                            3 => [
                                // -X (reversed U)
                                (p(sv, ii + ww, jj), [0.0, 0.0]),
                                (p(sv, ii + ww, jj + hh), [0.0, hh as f32]),
                                (p(sv, ii, jj + hh), [ww as f32, hh as f32]),
                                (p(sv, ii, jj), [ww as f32, 0.0]),
                            ],
                            4 => [
                                // +Z
                                (p(sv, ii, jj), [0.0, 0.0]),
                                (p(sv, ii + ww, jj), [ww as f32, 0.0]),
                                (p(sv, ii + ww, jj + hh), [ww as f32, hh as f32]),
                                (p(sv, ii, jj + hh), [0.0, hh as f32]),
                            ],
                            _ => [
                                // -Z (reversed U)
                                (p(sv, ii + ww, jj), [0.0, 0.0]),
                                (p(sv, ii, jj), [ww as f32, 0.0]),
                                (p(sv, ii, jj + hh), [ww as f32, hh as f32]),
                                (p(sv, ii + ww, jj + hh), [0.0, hh as f32]),
                            ],
                        };

                        let tbn = FACE_TBN[*face_index];
                        let sky_n = sky_l / 15.0;
                        let block_n = block_l / 15.0;

                        // Two triangles: (0,1,2) (0,2,3)
                        for vi in [0, 1, 2, 0, 2, 3] {
                            let (pos, uv) = &corners[vi];
                            positions.extend_from_slice(pos);
                            uvls.extend_from_slice(&[uv[0], uv[1], tile_idx, *light]);
                            normals.extend_from_slice(&[tbn[0], tbn[1], tbn[2]]);
                            tangents.extend_from_slice(&[tbn[3], tbn[4], tbn[5], tbn[6]]);
                            lights.extend_from_slice(&[sky_n, block_n]);
                        }
                    }
                }
            }
        }

        let vertex_count = (positions.len() / 3) as u32;
        ChunkMesh {
            positions,
            normals,
            uvls,
            tangents,
            lights,
            vertex_count,
        }
    }
}
